keps/sig-storage/20190408-volume-scheduling-limits.md

```diff
@@ -0,0 +1,242 @@
+  * If any of them is enabled together with `MaxCSIVolumeCountPred`, the deprecated predicate will do nothing (`MaxCSIVolumeCountPred` does the job of counting both in-tree and CSI volumes).
+  * The deprecated predicates will do any useful work only when `MaxCSIVolumeCountPred` predicate is disabled.
+  * This way, we save CPU by running only one volume limit predicate during deprecation period.
+
+
+### New API
+
+CSINode gets `Status` struct with `Allocatable`, holding limit of volumes for each volume plugin and CSI driver that can be scheduled to the node.
+
+```go
+type CSINode struct {
+	...
+
+	// spec is the status of CSINode
+	Status CSINodeStatus `json:"status" protobuf:"bytes,3,opt,name=status"`
+}
+
+// VolumeLimits is map CSI driver name -> maximum count of volumes for the driver on the node.
+// For in-tree volume plugins, name of corresponding CSI driver is used.
+// Value can be either:
+// - Positive integer: that's the volume limit.
+// - Zero: such volume cannot be used on the node.
+// - key is missing in VolumeLimits: there is no volume limit, i.e. any number of volumes can be used on the node.
+type VolumeLimits map[string]int64
+
+// CSINodeStatus holds information about the status of all CSI drivers installed on a node
+type CSINodeStatus struct {
+	// allocatable is a list of volume limits for each volume plugin and CSI driver on the node.
+	// +patchMergeKey=name
+	// +patchStrategy=merge
+	Allocatable []VolumeLimits `json:"allocatable" patchStrategy:"merge" patchMergeKey:"name" protobuf:"bytes,1,rep,name=allocatable"`
+}
+```
+
+### User Stories
+
+
+### Implementation Details/Notes/Constraints
+
+[CSI migration library](https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/csi-translation-lib) is used to find CSI driver name for in-tree volume plugins + its `VolumeHandle`. This CSI driver name is used as key in `CSINode.status.allocatable` list. The `VolumeHandle` is unique for each volume and will be used to de-duplicate volumes used by multiple pods on the same node.
+
+### Risks and Mitigations
+
+* This KEP depends on [CSI migration library](https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/csi-translation-lib). It can happen that CSI migration is redesigned / cancelled.
+  * Countermeasure: [CSI migration](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/20190129-csi-migration.md) and this KEP should graduate together.
+
+* This KEP depends on [CSI migration library](https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/csi-translation-lib) ability to handle in-line in-tree volumes. Scheduler will need to get CSI driver name + `VolumeHandle` from them to count them towards the limit. 
+
+## Design Details
+
+Existing feature gate `AttachVolumeLimit` will be re-used for implementation of this KEP. The feature is already beta and is enabled by default.
```

Comment:
Can you add a section about how backwards compatibility with the old beta implementation will work?

Reply:
Added to Design Details.